plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,304 @@
+    super.deactivated(workingCopy);
+    activated = false;
+
+    if (environmentTab != null) {
+      // Unfortunately, "EnvironmentTab" overrides "activated()" not to call "initializeFrom()".
+      // (Calling "initializeFrom()" when "activated()" is the default implementation of the base
+      // class retained for backward compatibility.) We needed to call it on behalf of
+      // "EnvironmentTab" to re-initialize its UI with the changes made here.
+      environmentTab.initializeFrom(workingCopy);
+    }
+  }
+
+  @Override
+  public void initializeFrom(ILaunchConfiguration configuration) {
+    accountEmailModel = getAttribute(configuration, ATTRIBUTE_ACCOUNT_EMAIL, ""); //$NON-NLS-1$
+
+    Map<String, String> environmentMap = getEnvironmentMap(configuration);
+    gcpProjectIdModel = Strings.nullToEmpty(environmentMap.get(PROJECT_ID_ENVIRONMENT_VARIABLE));
+    String serviceKey = Strings.nullToEmpty(environmentMap.get(SERVICE_KEY_ENVIRONMENT_VARIABLE));
+
+    initializingUiValues = true;
+    // Selecting an account loads projects into the project selector synchronously (via a listener).
+    accountSelector.selectAccount(accountEmailModel);
+    projectSelector.selectProjectId(gcpProjectIdModel);
+    serviceKeyInput.setText(serviceKey);
+    initializingUiValues = false;
+  }
+
+  @Override
+  public void performApply(ILaunchConfigurationWorkingCopy configuration) {
+    if (!activated) {  // to avoid https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2568#discussion_r150128582
+      return;
+    }
+
+    configuration.setAttribute(ATTRIBUTE_ACCOUNT_EMAIL, accountEmailModel);
+
+    Map<String, String> environmentMap = getEnvironmentMap(configuration);
+    if (!gcpProjectIdModel.isEmpty()) {
+      environmentMap.put(PROJECT_ID_ENVIRONMENT_VARIABLE, gcpProjectIdModel);
+    } else {
+      environmentMap.remove(PROJECT_ID_ENVIRONMENT_VARIABLE);
+    }
+    if (!serviceKeyInput.getText().isEmpty()) {
+      environmentMap.put(SERVICE_KEY_ENVIRONMENT_VARIABLE, serviceKeyInput.getText());
+    } else {
+      environmentMap.remove(SERVICE_KEY_ENVIRONMENT_VARIABLE);
+    }
+    configuration.setAttribute(ILaunchManager.ATTR_ENVIRONMENT_VARIABLES, environmentMap);
+  }
+
+  @Override
+  public String getName() {
+    return Messages.getString("gcp.emulation.tab.name"); //$NON-NLS-1$
+  }
+
+  @Override
+  public Image getImage() {
+    return gcpIcon;
+  }
+
+  @VisibleForTesting
+  static String getAttribute(ILaunchConfiguration configuration,
+      String attribute, String defaultValue) {
+    try {
+      return configuration.getAttribute(attribute, defaultValue);
+    } catch (CoreException e) {
+      logger.log(Level.WARNING, "Can't get value from launch configuration.", e); //$NON-NLS-1$
+      return defaultValue;
+    }
+  }
+
+  @VisibleForTesting
+  static Map<String, String> getEnvironmentMap(ILaunchConfiguration configuration) {
+    Map<String, String> emptyMap = new HashMap<>();  // should be mutable
```

Comment:
This comment is unexpected?  An explanation would be useful

Reply:
Wanted to signify we shouldn't use `Collections.emptyMap()` which is immutable. I'll update the comment.